File: 2015/rust/src/day4.rs
use md5::{Digest, Md5};
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::time::Instant;

pub fn day4(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 4: The Ideal Stocking Stuffer");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    let start = Instant::now();
    let sol1 = part1(&input)?;
    println!("\tPart 1: {} ({:?})", sol1, start.elapsed());
    let second = Instant::now();
    println!(
        "\tPart 2: {} ({:?})",
        part2(&input, sol1)?,
        second.elapsed()
    );
    println!("\t\t Completed in {:?}", start.elapsed());
    Ok(())
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    let mut i = 1;
    let mut hasher = Md5::new();
    loop {
        let mut prefix = input.to_owned();
        prefix += &i.to_string();
        hasher.update(prefix.as_bytes());
        let hash = hasher.finalize_reset();

        if hash[0] == 0 && hash[1] == 0 && hash[2] & 0xf0 == 0 {
            return Ok(i);
        }
        i += 1;
    }
}

fn part2(input: &str, start: i32) -> Result<i32, Box<dyn Error>> {
    let mut i = start;
    let mut hasher = Md5::new();
    loop {
        let mut prefix = input.to_owned();
        prefix += &i.to_string();
        hasher.update(prefix.as_bytes());
        let hash = hasher.finalize_reset();

        if hash[0] == 0 && hash[1] == 0 && hash[2] == 0 {
            return Ok(i);
        }
        i += 1;
    }
}

File: 2015/c/src/day4.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/md5.h>
#include "lib.h"

#define MAX_INT_CHARS 10
static int part1(const char *input, size_t input_len)
{
    char key[input_len + MAX_INT_CHARS];
    size_t key_len = strlen(input);
    strcpy(key, input);

    int i = 1;
    while (true)
    {
        memset(key + key_len, 0, input_len + MAX_INT_CHARS - key_len);
        sprintf(key + key_len, "%d", i);
        unsigned char *sum = MD5((unsigned char *)key, strlen(key), NULL);
        if (sum[0] == 0 && sum[1] == 0 && (sum[2] & 0xf0) == 0)
            return i;

        i++;
    }
}

static int part2(const char *input, size_t input_len, int start)
{
    char key[input_len + MAX_INT_CHARS];
    size_t key_len = strlen(input);
    strcpy(key, input);

    int i = start;
    while (true)
    {
        memset(key + key_len, 0, input_len + MAX_INT_CHARS - key_len);
        sprintf(key + key_len, "%d", i);
        unsigned char *sum = MD5((unsigned char *)key, strlen(key), NULL);
        if (sum[0] == 0 && sum[1] == 0 && sum[2] == 0)
            return i;

        i++;
    }
}

int day4(const char *filename)
{
    printf("Day 4: The Ideal Stocking Stuffer\n");

    char *input;
    int filesize = read_file_to_buffer(&input, filename);
    if (filesize < 0)
        return -1;

    int rval = part1(input, filesize);
    if (rval < 0)
        return rval;
    printf("\tPart 1: %d\n", rval);

    rval = part2(input, filesize, rval);
    if (rval < 0)
        return rval;
    printf("\tPart 2: %d\n", rval);

    free(input);

    return 0;
}
